ionoscloud/data_source_k8s_clusters.go

```diff
@@ -0,0 +1,234 @@
+package ionoscloud
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/iancoleman/strcase"
+	ionoscloud "github.com/ionos-cloud/sdk-go/v6"
+	"github.com/ionos-cloud/terraform-provider-ionoscloud/v6/internal/uuidgen"
+	"github.com/ionos-cloud/terraform-provider-ionoscloud/v6/services"
+	"github.com/ionos-cloud/terraform-provider-ionoscloud/v6/utils"
+	"gopkg.in/yaml.v3"
+)
+
+func dataSourceK8sClusters() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceK8sReadClusters,
+		Schema: map[string]*schema.Schema{
+			"clusters": {
+				Type:        schema.TypeList,
+				Description: "List of clusters which match the filtering criteria.",
+				Computed:    true,
+				Elem:        &schema.Resource{Schema: dataSourceK8sClusterSchema()},
+			},
+			"entries": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"filter": dataSourceFiltersSchema(),
+		},
+		Timeouts: &resourceDefaultTimeouts,
+	}
+}
+
+func dataSourceK8sReadClusters(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+
+	// strcase.ToLowerCamel doesn't produce the correct format for these keys, provide them directly (https://github.com/iancoleman/strcase/issues/51)
+	filterKeys := map[string]string{
+		"k8s_version": "k8sVersion",
+	}
+
+	client := meta.(services.SdkBundle).CloudApiClient
+	req := client.KubernetesApi.K8sGet(ctx).Depth(1)
+
+	filters, filtersOk := d.GetOk("filter")
+	if !filtersOk {
+		return diag.FromErr(errors.New("please provide filters for data source lookup"))
+	}
+
+	for _, v := range filters.(*schema.Set).List() {
+		filter := v.(map[string]any)
+		key := filter["name"].(string)
+		value := filter["value"].(string)
+		if v, ok := filterKeys[key]; ok {
+			key = v
+		} else {
+			key = strcase.ToLowerCamel(key)
+		}
+		req.Filter(key, value)
+	}
+
+	clusters, apiResponse, err := req.Execute()
+	logApiRequestTime(apiResponse)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("an error occurred while fetching k8s clusters: %w", err))
+	}
+	if clusters.Items != nil && len(*clusters.Items) == 0 {
+		return diag.FromErr(fmt.Errorf("no clusters match the specified filtering criteria"))
```

Comment:
can we add `filters` to the err message?

Reply:
the filters that were used are now displayed in the error message ([f6d84ca](https://github.com/ionos-cloud/terraform-provider-ionoscloud/pull/519/commits/f6d84ca77aa5f5db67d18e25d54353ec8a480c4a))